### PR TITLE
Add eid types for the eid server

### PR DESCRIPTION
**Pull request by m273d15**
This PR (for issue #51 ) replaces #65 .
* Replaced assert by assertEqual
* Replaced try catch by assertRaises
* Added blank lines as defined in PEP8
* Added newline if line is longer than 120 characters
* Sorted commits by types
* Moved all types and tests from the directory `boilerplate` to `eid_server`
Created by @m273d15 and @larissazech

### CODE REVIEW

File: run_ci.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+js_test=$(docker ps -a -f exited=1 -q | wc -l | tr -d ' ')
+docker-compose exec -T api sh -c 'python manage.py test; (exit $?)'
+py_test=$?
+echo $js_test
+echo $py_test
+
+docker-compose down
+
+if [[ $js_test == 0 && $py_test == 0 ]]; then
+exit 0
```
Comment (nils-wisiol) [inline run_ci.sh]: indentation

File: api/boilerplate/eid_types/sessionid.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class SessionID:
+
+    def __init__(self):
+        self.__sessionid = ""
+
+    @property
+    def id(self):
+        i = 16
```
Comment (nils-wisiol) [inline api/boilerplate/eid_types/sessionid.py]: What does this do? Maybe add some docs.

File: api/boilerplate/eid_types/sessionid.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        return self.__sessionid
+
+    @id.setter
+    def id(self, value):
+        if len(value) < 32 or len(value) > 64:
+            return False
+
+        for c in value:
+            if (c not in list(range(ord('0'), ord('9') + 1)) 
+            + list(range(ord('a'), ord('f') + 1))):
```
Comment (nils-wisiol) [inline api/boilerplate/eid_types/sessionid.py]: indentation

File: api/boilerplate/eid_types/hexbinary.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        self.hex_string = hex_string
+
+    @property
+    def hex_string(self):
+        return self._hex_string
+
+    @hex_string.setter
+    def hex_string(self, value):
+
+        pattern = "^([A-F,a-f,0-9]{2})*$"
```
Comment (nils-wisiol) [inline api/boilerplate/eid_types/hexbinary.py]: please use python built-in regex

File: api/boilerplate/eid_types/hexbinary.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        check_is_string_matching("Hex string should match the regex " + pattern, "hex_string",
+        value, pattern)
```
Comment (nils-wisiol) [inline api/boilerplate/eid_types/hexbinary.py]: indentation

**Review by nils-wisiol — CHANGES_REQUESTED** [review]
- Please fix indentation, in particular after line breaks
- please use python regex rather than run time-compiled expressions

**Comment by larissazech** [comment]
@nils-wisiol, @m273d15 and I fixed the indentation and the regex and rebased to master. Can we merge?

File: api/eid_server/eid_types/sessionid.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import numpy as np
```
Comment (nils-wisiol) [inline api/eid_server/eid_types/sessionid.py]: unused import

File: api/eid_server/eid_types/sessionid.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import binascii
+
+
+class SessionID:
+
+    def __init__(self):
+        self.__sessionid = ""
+
+    @property
+    def id(self):
```
Comment (nils-wisiol) [inline api/eid_server/eid_types/sessionid.py]: please add some docs here

File: api/eid_server/eid_types/sessionid.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        i = 16
+        while len(self.__sessionid) < 32:
+            a = random.bytes(i)
```
Comment (nils-wisiol) [inline api/eid_server/eid_types/sessionid.py]: numpy.random is not a cryptographically safe way to generate random numbers

Comment (nils-wisiol) [reply api/eid_server/eid_types/sessionid.py]: why don't you just generate 16 bytes of random data?

Also, you could use [random.choice](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.random.choice.html#numpy.random.choice)? (After seeding it with crypto-safe random numbers.)

**Review by nils-wisiol — COMMENTED** [review]
- generation of session ids needs refactoring
- [avoid `import *`](https://stackoverflow.com/a/2386740) (the test's `__init__` file is an exception here)
- scipy appears to be not needed, don't install it

**Comment by larissazech** [comment]
Taken from the eID Server specification:

> The type SessionType SHALL be used to uniquely identify a request to the eID-Interface. The eID-Server MUST generate values for the ID element to identify different requests for Online-Authentication. The value of the element ID MUST be random and at least 32 characters long in hexadecimal representation so that it's not easy to guess. The software component generating the random numbers used as IDs SHALL fulfill the requirements described in [TR-03116-4] Section 6.2: Zufallszahlen and the eID-Service MUST ensure that no active sessions using identical IDs exist in the same context.The eService SHALL only use valid session IDs for calls of the function getResult.

I would suggest to leave the generation of a random number to the component instantiating a SessionType and to only check for the 16Bits HexBinary (as described in the xsd) here. 

@m273d15 @Armagetron Agreed?

File: api/eid_server/eid_types/sessionid.py [hunk 9]
```diff
@@ -1,32 +1,18 @@
...
-            return False
-
-        for c in value:
-            if (c not in list(range(ord('0'), ord('9') + 1)) 
-                    + list(range(ord('a'), ord('f') + 1))):
-                return False
-
-        self.__sessionid = value
-        return True
+        check("id", value, HexBinary16)
```
Comment (larissazech) [inline api/eid_server/eid_types/sessionid.py]: I refactored as suggested. For being able to rewind the changes, I pushed a whole new commit. If the changes are approved, I will squash it to Commit 702a6c493437e41deb6a27b1cde03bf53c6f47c4 .

**Comment by larissazech** [comment]
We discussed the SessionID-issue internally and agreed upon the suggested solution. Commit history has been cleaned. Can we merge?

File: api/eid_server/tests/personaldata.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    VALID_DOCUMENT_TYPE = DocumentType("DE")
+    VALID_ISSUING_STATE = ICAOCountry("D")
+    VALID_DATE_OF_EXPIRY = datetime.today()
+    VALID_GIVEN_NAMES = "Max"
+    VALID_FAMILY_NAMES = "Mustermann"
+    VALID_ARTISTIC_NAME = "Mozart"
+    VALID_ACADEMIC_TITLE = "Dr"
+    VALID_DATE_OF_BIRTH = GeneralDate("19900124")
+    VALID_PLACE_OF_BIRTH = GeneralPlace(Place("Berlin", ICAOCountry("D")), 
+    "Berlin, DE", "no place info")
```
Comment (nils-wisiol) [inline api/eid_server/tests/personaldata.py]: indentation

**Review by nils-wisiol — CHANGES_REQUESTED** [review]
The specification you quoted yourself explicitly states

> The eID-Server MUST generate values for the ID element

So, unless the other component is also part of the eId server, I'd say this needs to be done here.

Unfortunately, [this is not ready to merge](https://github.com/swp-fu-eid/eid-fu-swp/pull/70#pullrequestreview-61264027).

**Review by nils-wisiol — APPROVED** [review]

**Comment by nils-wisiol** [comment]
Let's merge this once we need it

**Pull request merged**